cmd/tealdbg/localLedger.go

```diff
@@ -18,103 +18,103 @@ package main
+	for _, addr := range txn.Txn.Accounts {
+		accounts = append(accounts, addr)
+	}
 
-func (l *localLedger) AppGlobalState(appIdx basics.AppIndex) (basics.TealKeyValue, error) {
-	var ap basics.AppParams
+	appIdx := txn.Txn.ApplicationID
 	if appIdx == 0 {
-		if l.groupIndex >= len(l.txnGroup) {
-			return basics.TealKeyValue{}, fmt.Errorf("can't resolve application %d", appIdx)
-		}
-		appIdx = l.txnGroup[l.groupIndex].Txn.ApplicationID
+		// presumably this is app create transaction, initialize with some value
+		appIdx = defaultNewAppIdx
 	}
 
-	allowed := false
-	if appIdx == l.txnGroup[l.groupIndex].Txn.ApplicationID {
-		allowed = true
-	} else {
-		for _, faIdx := range l.txnGroup[l.groupIndex].Txn.ForeignApps {
-			if appIdx == faIdx {
-				allowed = true
-				break
-			}
-		}
+	apps := []basics.AppIndex{appIdx}
+	for _, aid := range txn.Txn.ForeignApps {
+		apps = append(apps, aid)
 	}
-	if !allowed {
-		return basics.TealKeyValue{}, fmt.Errorf("access to the app forbidden %d", appIdx)
+
+	ba := &balancesAdapter{
+		balances:   balances,
+		txnGroup:   txnGroup,
+		groupIndex: groupIndex,
+		proto:      proto,
+		round:      round,
 	}
 
-	for _, br := range l.balances {
-		var ok bool
-		ap, ok = br.AppParams[appIdx]
-		if ok {
-			return ap.GlobalState, nil
-		}
+	return ledger.MakeDebugAppLedger(ba, accounts, apps, appIdx)
+}
 
+func (ba *balancesAdapter) Get(addr basics.Address, withPendingRewards bool) (basics.BalanceRecord, error) {
+	br, ok := ba.balances[addr]
+	if !ok {
+		return basics.BalanceRecord{}, fmt.Errorf("no such address %s", addr.String())
 	}
-	return basics.TealKeyValue{}, fmt.Errorf("no such application %d", appIdx)
+	return basics.BalanceRecord{Addr: addr, AccountData: br}, nil
 }
 
-func (l *localLedger) AppLocalState(addr basics.Address, appIdx basics.AppIndex) (basics.TealKeyValue, error) {
-	if appIdx == 0 {
-		if l.groupIndex >= len(l.txnGroup) {
-			return basics.TealKeyValue{}, fmt.Errorf("can't resolve application %d", appIdx)
+func (ba *balancesAdapter) Round() basics.Round {
+	return basics.Round(ba.round)
+}
+
+func (ba *balancesAdapter) GetAssetCreator(assetIdx basics.AssetIndex) (basics.Address, bool, error) {
+	for addr, br := range ba.balances {
+		if _, ok := br.AssetParams[assetIdx]; ok {
+			return addr, true, nil
 		}
-		appIdx = l.txnGroup[l.groupIndex].Txn.ApplicationID
 	}
+	return basics.Address{}, false, fmt.Errorf("not found")
+}
 
-	br, ok := l.balances[addr]
-	if !ok {
-		return basics.TealKeyValue{}, fmt.Errorf("no such address %s", addr.String())
-	}
-	ls, ok := br.AppLocalStates[appIdx]
-	if !ok {
-		return basics.TealKeyValue{}, fmt.Errorf("no local state for application %d", appIdx)
+func (ba *balancesAdapter) GetAppCreator(appIdx basics.AppIndex) (basics.Address, bool, error) {
+	for addr, br := range ba.balances {
+		if _, ok := br.AppParams[appIdx]; ok {
+			return addr, true, nil
+		}
 	}
-	return ls.KeyValue, nil
+	return basics.Address{}, false, fmt.Errorf("not found")
```

Comment:
In general this function won't return a non-nil `error` just because the app wasn't found -- it should probably be returning `basics.Address{}, false, nil` here. Same for `GetAssetCreator`.